### PR TITLE
New way of handling emblems with theming support (boo#1109382)

### PR DESCRIPTION
**Pull request by hellcp**

### CODE REVIEW

File: src/YQIconPool.cc [hunk 1]
```diff
@@ -38,106 +38,64 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
-QPixmap YQIconPool::stepCurrent()		{ return iconPool()->cachedIcon( step_current_xpm );		}
-QPixmap YQIconPool::stepToDo()			{ return iconPool()->cachedIcon( step_todo_xpm );		}
-QPixmap YQIconPool::stepDone()			{ return iconPool()->cachedIcon( step_done_xpm );		}
-
-QPixmap YQIconPool::tabRemove()			{ return iconPool()->cachedIcon( tab_remove_xpm );		}
-QPixmap YQIconPool::arrowLeft()			{ return iconPool()->cachedIcon( arrow_left_xpm );		}
-QPixmap YQIconPool::arrowRight()		{ return iconPool()->cachedIcon( arrow_right_xpm );		}
-QPixmap YQIconPool::arrowUp()			{ return iconPool()->cachedIcon( arrow_up_xpm );		}
-QPixmap YQIconPool::arrowDown()			{ return iconPool()->cachedIcon( arrow_down_xpm );		}
+QPixmap YQIconPool::pkgTaboo()			{ return iconPool()->cachedIcon( "yui-taboo" );				}
```
Comment (lslezak) [inline src/YQIconPool.cc]: Looks like there are some extra white space characters (space before tab) in this block.

Comment (hellcp) [reply src/YQIconPool.cc]: Can't see those locally, I'll guess Github is displaying it wrong

Comment (lslezak) [reply src/YQIconPool.cc]: Yes, you are right, the GH tab rendering is broken...

**Comment by shundhammer** [comment]
This PR is technically correct AFAICS.

But I have a problem with the overall approach: Icons are no longer compiled into the application, they are **always** loaded from separate files.

While this is okay for many things (control center etc.) where the icons are purely decorative, that means that now also the YQPkgSelector also depends on those external files.

This is bad: This means that if somebody managed to wreck his icon directory, he can no longer use the package manager. This was the main reason why those icons were compiled in.

I can understand that it is desirable from your POV to have themeable icons. But it is also important to always have fallback icons for those package states; we really **need** icons that are compiled into the source.

So I propose to do this:

- Use the Qt resource file system to pack appropriate icons (at least for the package status icons) into a Qt resource file that is liked to the Qt UI.

- In the icon loader, try to load icons from external file; if that fails, fall back to the internal ones.

To do that, simply use a path ":/icons/foo.svg"; the colon is what marks those files as coming from a resource file.

Example how to use this:

https://github.com/shundhammer/qdirstat/tree/master/src

(Using the icons in the Qt Designer .ui files with their ":/icons/*" paths)

The challenge here is how to get this into the CMake build environment; this is the part that I don't know. But there should be documentation about this on the web. My QDirStat does it with Qt's _qmake_.

**Comment by shundhammer** [comment]
More info on Qt resource files:

http://doc.qt.io/qt-5/resources.html

**Comment by shundhammer** [comment]
CMake and Qt resource files:

http://doc.qt.io/qt-5/cmake-manual.html

**Comment by hellcp** [comment]
I like that approach, I will look into it, thanks :D

**Comment by hellcp** [comment]
Hmmm, for some weird reason `qrc:///` didn't work but `:/` did, anyway, spent way too much time on this :P

File: src/YQIconPool.cc [hunk 2]
```diff
@@ -162,16 +119,27 @@ YQIconPool::~YQIconPool()
...
+    QPixmap iconPixmap = _iconCache[ icon_name ];
+    Q_INIT_RESOURCE(icons);
 
-    if ( iconPtr.isNull() )
+    if ( iconPixmap.isNull() )
     {
-	iconPtr = QPixmap( xpm_data );
-	_iconCache.insert( xpm_data, iconPtr );
+      if ( QIcon::hasThemeIcon(icon_name) )
+      {
```
Comment (shundhammer) [inline src/YQIconPool.cc]: Just a minor thing: Indentation in YaST C++ code is 4 blanks, not 2. Please reindent that function.

**Comment by shundhammer** [comment]
Just that minor indentation thing, otherwise LGTM now once Travis is happy.

File: src/YQPackageSelectorHelp.cc [hunk 3]
```diff
@@ -246,10 +246,10 @@ YQPackageSelector::symHelp( const QString & imgFileName,
 			    const QString & summary,
 			    const QString & explanation		)
 {
-    //FIXME: This should be HELP_ICONS_DIR but is set wrong by libyui
-    QString imgPath = "/usr/share/YaST2/images";
+    //FIXME: This should be specified by Qt on theme by theme basis, right now full path
+    QString imgPath = "/usr/share/icons/hicolor/16x16/emblems";
     QString html = "<tr valign='top'>";
-    html += "<td><img src=\"" + imgPath + "/" + imgFileName + "\"></td>"
+    html += "<td><img src=\"" + imgPath + "/yast-emblem-" + imgFileName + ".svg\"></td>"
```
Comment (mvidner) [inline src/YQPackageSelectorHelp.cc]: I see no "yast-emblem-*" files anywhere, this seems to be wrong.

**Comment by hellcp** [comment]
This approach makes it possible to use already existing breeze icons for package management:

screenshot from 2018-10-27 14-58-18

screenshot from 2018-10-27 14-58-33

Help is now also themed :D

**Comment by mvidner** [comment]
I've fixed the changelog in c03570aadbd6c75ee816c606cbf7eee189e2fdad, and updated the ci.suse.de workers to libyui-rake-0.1.12 which skips the license check for *.qrc.

**Pull request merged**